### PR TITLE
gcc-4.8: Fix default include path for system headers

### PR DESCRIPTION
**Pull request by stuarteberg**
There is a critical problem with conda's gcc package, which this PR addresses.

**tl;dr**: If we install gcc to `$PREFIX`, then gcc automatically considers `$PREFIX/include` to be a "system" header path.  That can mess up the include order.  This PR installs gcc to a special place (`$PREFIX/gcc`) to avoid the problem.

**Background:**

As we all know, when gcc searches for a header file, it searches any paths the user provided via `-I`, and then gcc also searches an internal list of "system" directories.  In general, the user's directories are searched in the same order as the user's `-I` arguments. But if the user supplies a `-I` argument that happens to *already* be on gcc's internal "system" include path, then that particular `-I` argument is ignored -- it's a system directory, and system directories are always searched last.  (See gcc docs: https://gcc.gnu.org/onlinedocs/cpp/Search-Path.html)

**The problem:**

When building gcc, the `--prefix` configure argument specifies the installation directory.  As it turns out, gcc **automatically adds** `$(prefix)/include` to its internal "system" include path. [1]  Since conda's gcc recipe `--prefix=$PREFIX`, the environment's `$PREFIX/include` directory is considered to be a *system* directory by gcc.

\[1\]: The [gcc configuration docs](https://gcc.gnu.org/install/configure.html) say:
>Both the local-prefix include directory and the GCC-prefix include directory are part of GCC's “system include” directories.

**Why is that bad?**

Now the user can't control the include order properly.  In my case, an old version of a library (say, hdf5) happened to be installed in `/usr/local/include`.  Ordinarily, it shouldn't cause any problems because I passed `-I $PREFIX/include` to the compiler, so my hdf5 headers in `$PREFIX/include` should be found first and used in my program.  But since gcc thinks `$PREFIX/include` is a system header, my `-I` argument is ignored, and that directory is searched last.  The outdated hdf5 headers in `/usr/local/include` were found first, and my recipe didn't compile properly.

**The Fix:**

gcc can be configured to install its files in custom install locations for `lib`, `bin`, `share`, etc., but apparently `include` cannot be separately controlled.  The only way to change the installation `include` directory is to change the entire install prefix.  In this PR, instead of `--prefix=$PREFIX`, I use `--prefix=$PREFIX/gcc`.  Then, to maintain compatibility with existing recipes, I also provide extra configure args to make sure that `bin`, `lib`, etc. stay in the same locations they had previously.  Only the gcc headers have moved.

**Addendum:**

While I was at it, I fixed some other minor issues with the `gcc` package.  I could have opened separate PRs for those, but in the past it's been easier to consider multiple changes at once (gcc is tedious to compile).

### CODE REVIEW

**Comment by stuarteberg** [comment]
BTW, You can get a pre-compiled version of this PR with my channel:

    conda install -c stuarteberg gcc=4.8.5

**Comment by stuarteberg** [comment]
PS -- The easiest way to check the gcc include order is with `gcc -v`:

With the old gcc package (comments added):
```
$ echo | gcc -v -x c -
...
...
#include "..." search starts here:
#include <...> search starts here:
 /miniconda/envs/myenv/lib/gcc/x86_64-unknown-linux-gnu/4.8.5/include
 /usr/local/include
 /miniconda/envs/myenv/include #<-- *** Whoops, that's not supposed to be a system path ***
 /miniconda/envs/myenv/lib/gcc/x86_64-unknown-linux-gnu/4.8.5/include-fixed
 /usr/include
End of search list.
...
```

With this PR's gcc package (comments added):
```
$ echo | gcc -v -x c -
...
...
#include "..." search starts here:
#include <...> search starts here:
 /miniconda/envs/myenv/lib/gcc/x86_64-unknown-linux-gnu/4.8.5/include
 /usr/local/include
 /miniconda/envs/myenv/gcc/include #<-- *** It's okay if this is a system path ***
 /miniconda/envs/myenv/lib/gcc/x86_64-unknown-linux-gnu/4.8.5/include-fixed
 /usr/include
End of search list.
...
```

**Comment by asmeurer** [comment]
Is this ready to merge and test?

**Comment by stuarteberg** [comment]
>Is this ready to merge and test? 
I've done some testing, but now I'm building a couple larger packages just to be sure (e.g. boost).  I'll know in a few minutes if there were any surprises.

BTW, I had trouble testing this on OS X because of #302, and a few other issues related to the linker.  I had to make a few additional commits (see the commit log).  Without those commits, I don't understand how you got the original package to compile on OSX.  Clearly I might be missing something.

Anyway, the current recipe settings seem to work just fine, but I would appreciate a quick review.  (It should be easier to grok if you review the commits one at a time.)

**Comment by stuarteberg** [comment]
OK, I can build boost-1.55.0 without problems on CentOS 5.11, Ubuntu 14.04, and OS X 10.10.  (I did not test any 32-bit systems.)

Meanwhile, shellcheck found one tiny quote issue, which I've now pushed (via a rebase).

**Comment by stuarteberg** [comment]
This PR is ready, as far as I know.

**Comment by asmeurer** [comment]
OK, I'll try building it and let you know if I come across any issues..

**Comment by stuarteberg** [comment]
>OK, I'll try building it and let you know if I come across any issues.. 

@asmeurer, have you gotten a chance to test the new gcc package yet?  Any idea when you'll be pushing your [new build](https://binstar.org/asmeurer/gcc/4.8.5/download/linux-32/gcc-4.8.5-2.tar.bz2) over to the "official" anaconda default channel?

**Comment by stuarteberg** [comment]
Just a quick ping to the participants of #279: The gcc recipe was updated again last week, via this PR.  If you copied the gcc package to your own channel, you should eventually upgrade to the new version.  BTW @frol, I assume these changes should also be made to the `gcc-5` recipe, too.

At the moment, the new version is not yet pushed to the default anaconda channel, but for now you can use @asmeurer's build: `conda install -c asmeurer gcc=4.8.5`.  (The new version has `build_number=2`.)

@jakirkham, @connie, @orodbhen, @mjuric, @frol, @izaid, @dfroger

**Comment by frol** [comment]
@asmeurer BTW, my PR #373 with `make strip-install` hasn't still made into the master. Is there any (design?) problem?

@stuarteberg Sure I will update `gcc-5` recipe. While we are here again, I have revisited post-link and pre-unlink scripts and the logic of `if [ "$(uname)" == "Darwin" ]; then exit ; fi` (https://github.com/stuarteberg/conda-recipes/blob/ed718bb5f44f53e63a7947842b60244dac1ab6db/gcc-4.8/post-link.sh#L3) seems wrong-targeted to me. IMHO, it would be better to use `if [ "$(uname)" != "Linux" ]; then exit ; fi`

**Comment by stuarteberg** [comment]
>I have revisited post-link and pre-unlink scripts and the logic of `if [ "$(uname)" == "Darwin" ]; then exit ; fi` seems wrong-targeted to me. IMHO, it would be better to use `if [ "$(uname)" != "Linux" ]; then exit ; fi`

The logic in `post-link.sh` could (in theory) be applied to any unix-ish OS, including Darwin.  But we happen to know that Darwin doesn't need the fix, so we figured it would be better just to skip post-link entirely in that case.  (It wouldn't do any harm to run it on Darwin, but it seems faster/safer just to skip post-link since we know it isn't needed.)

But someday when conda supports other non-Linux OS's (I dunno... Solaris?), then we probably wouldn't want to skip post-link.  For now, a special case for Darwin is fine.
